### PR TITLE
Create react-select package

### PR DESCRIPTION
**Pull request by smhigley**
#### Pull request checklist

- [x] Addresses an existing issue: #20935
- [ ] ~Include a change request file using `$ yarn change`~ (not included, private package)

#### Description of changes

Adds a `react-select` package and basic component structure and stories. Does not include styles.

### CODE REVIEW

**Comment by fabricteam** [comment]
## 📊 Bundle size report

<details>
<summary>Unchanged fixtures</summary>

| Package & Exports                                                                                                                        |        Size (minified/GZIP) |
| ---------------------------------------------------------------------------------------------------------------------------------------- | --------------------------: |
| <samp>react-make-styles</samp> <br /> <abbr title='bundle-size/makeStaticStyles-runtime.fixture.js'>makeStaticStyles (runtime)</abbr>    |  `7.602 kB`<br />`3.325 kB` |
| <samp>react-make-styles</samp> <br /> <abbr title='bundle-size/makeStyles-runtime.fixture.js'>makeStyles + mergeClasses (runtime)</abbr> | `22.006 kB`<br />`8.311 kB` |
| <samp>react-make-styles</samp> <br /> <abbr title='bundle-size/makeStyles.fixture.js'>makeStyles + mergeClasses (build time)</abbr>      |   `2.689 kB`<br />`1.23 kB` |
| <samp>react-theme</samp> <br /> <abbr title='bundle-size/TeamsAllThemes.fixture.js'>Teams: all themes</abbr>                             | `29.506 kB`<br />`6.512 kB` |
| <samp>react-theme</samp> <br /> <abbr title='bundle-size/TeamsLightTheme.fixture.js'>Teams: Light theme</abbr>                           | `18.517 kB`<br />`5.237 kB` |
| <samp>react-utilities</samp> <br /> <abbr title='bundle-size/SSRProvider.fixture.js'>SSRProvider</abbr>                                  |        `189 B`<br />`161 B` |

</details>
<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/3defe8482167a065a9ee9653fd0c34d5d3ecf055'>3defe8482167a065a9ee9653fd0c34d5d3ecf055</a></sub>
<div id="bundleSizeComment9423"/>

File: packages/react-select/.storybook/main.js
Comment (ling1726) [inline packages/react-select/.storybook/main.js:3]: you might wanna rebase master and run `yarn nx workspace-generator migrate-converged-pkg --name='@fluentui/select'` again since I the config here isn't entirely corret (missing rootMain spread on the root)

File: packages/react-select/src/components/Select/Select.stories.tsx
Comment (ling1726) [inline packages/react-select/src/components/Select/Select.stories.tsx:1]: This shouldn't be needed 

```suggestion
```

File: packages/react-select/src/components/Select/useSelectStyles.ts
Comment (ling1726) [inline packages/react-select/src/components/Select/useSelectStyles.ts:6]: We have proposed sharing common style mixins, but there's no conclusion on that unfortunately :(

If these styles are the same as what is in `Text` then we should reevaluate this, because we shouldn't copy+paste styles

File: packages/react-select/src/components/Select/useSelectStyles.ts
Comment (ling1726) [inline packages/react-select/src/components/Select/useSelectStyles.ts:38]: There's no real reason to use more than one invocation of `makeStyles` you also end up calling less hooks which might be a good thing

File: packages/react-select/src/components/Select/useSelectStyles.ts
Comment (ling1726) [inline packages/react-select/src/components/Select/useSelectStyles.ts:78]: We intend of banning `background` in the future due to the shorthand expansion problem, more details in #20573

```suggestion
    backgroundColor: backgroundColors.filledLighter(theme),
```

File: packages/react-select/src/components/Select/useSelectStyles.ts
Comment (ling1726) [inline packages/react-select/src/components/Select/useSelectStyles.ts:56]: can we reuse the default focus outline styles as other components ? We can add a transparent outline there (it's not there currently), but I recently cleared up with Design that we can use it for windows HC

File: packages/react-select/tsconfig.json
Comment (ling1726) [inline packages/react-select/tsconfig.json:1]: you should definitely re-run the migration script I mentioned earlier since you're not using the new tsconfig setup here

File: packages/react-select/src/components/Select/useSelectStyles.ts
Comment (ling1726) [inline packages/react-select/src/components/Select/useSelectStyles.ts:94]: nit: this works too if you want to remove a line of code

```suggestion
  const { size = 'medium', appearance = 'outline', select: { disabled } } = state;
```

File: packages/react-select/src/components/Select/Select.stories.tsx
Comment (ling1726) [inline packages/react-select/src/components/Select/Select.stories.tsx:38]: It would be nice to have a controlled story that uses `onChange`

File: packages/react-select/src/components/Select/Select.stories.tsx
Comment (ling1726) [inline packages/react-select/src/components/Select/Select.stories.tsx:18]: Let's follow the same stories pattern we have in other components. This is primarily so that we can export all the examples to codesandbox cleanly

* one story per file
* each story only imports what can run on codesandbox
* each story is reexported by a main module for storybook

File: packages/react-select/src/components/Select/useSelect.ts
Comment (layershifter) [inline packages/react-select/src/components/Select/useSelect.ts:31]: As I see `disabled` is defined in `SelectProps`, is it intentional that it's not passed down to `select`?

If not:

```suggestion
    select: resolveShorthand(select, { defaultProps: { disabled }, required: true }),
```

Wider question: I would except that `Select` will follow "primary slot" pattern (sample implementation #20863). Do you want to address this in current PR or in a separate?

**Review by JustSlone — APPROVED** [review]
Approving codeowners change

File: packages/react-select/src/components/Select/useSelectStyles.ts
Comment (smhigley) [reply packages/react-select/src/components/Select/useSelectStyles.ts:78]: I'm actually writing a separate PR for the styles, so I might remove this entirely (I wrote them some time ago, the ones here are from before the shorthand discussion)

File: packages/react-select/src/components/Select/useSelect.ts
Comment (smhigley) [reply packages/react-select/src/components/Select/useSelect.ts:31]: I'd planned to update that in a separate PR (actually on my todo list)

File: packages/react-select/src/components/Select/useSelectStyles.ts
Comment (smhigley) [reply packages/react-select/src/components/Select/useSelectStyles.ts:6]: I agree, going to leave this convo to the styling PR to not overload this one (and removing all specific styles from this PR, since I was rewriting them anyway)

File: packages/react-select/src/components/Select/useSelect.ts
Comment (smhigley) [reply packages/react-select/src/components/Select/useSelect.ts]: I updated this PR to have `select` be the primary slot 👍

File: packages/react-select/src/components/Select/Select.types.ts
Comment (ling1726) [inline packages/react-select/src/components/Select/Select.types.ts:29]: wh did yo need to omit `size` ? is there a native size prop that's conflicting with the one for the component  ?

File: packages/react-select/src/components/Select/useSelect.ts
Comment (ling1726) [inline packages/react-select/src/components/Select/useSelect.ts:20]: I don't see anywhere else you are applying defaults

```suggestion
  const { select, icon, root, size = 'medium', appearance = 'outline', inline } = props;
```

File: packages/react-select/src/stories/SelectControlled.stories.tsx
Comment (ling1726) [inline packages/react-select/src/stories/SelectControlled.stories.tsx:5]: I think we also need an example with `onChange` since I guess the most common use case of a select is to know when someone has made a choice

File: packages/react-select/src/components/Select/Select.types.ts
Comment (smhigley) [reply packages/react-select/src/components/Select/Select.types.ts:29]: Yup! The native `size` attribute is explicitly not supported, and the spec goes over details on why :)

File: packages/react-select/src/stories/SelectControlled.stories.tsx
Comment (smhigley) [reply packages/react-select/src/stories/SelectControlled.stories.tsx:5]: The reason I didn't add it in this PR was b/c I was somewhat hoping to wait until the onChange RFC was merged. I do plan to include that at some point, though. I could update the Select epic with a checklist that includes updating the controlled example with the standardized `onChange` handler -- does that sound like a good approach?

Comment (ling1726) [reply packages/react-select/src/stories/SelectControlled.stories.tsx:5]: yes either that or a separate issue, it's quite important for this component

File: packages/react-select/src/components/Select/useSelect.ts
Comment (smhigley) [reply packages/react-select/src/components/Select/useSelect.ts:20]: I added defaults in the style PR, but I can go back and add them here too

File: packages/react-select/src/components/Select/useSelectStyles.ts
Comment (smhigley) [reply packages/react-select/src/components/Select/useSelectStyles.ts]: I actually removed this entirely when I took out the styles from this PR, but updated it in the styles PR

File: packages/react-select/src/components/Select/useSelectStyles.ts
Comment (smhigley) [reply packages/react-select/src/components/Select/useSelectStyles.ts]: Select and Input actually don't have the same focus styles as other components, so the default outline styles wouldn't apply. I've removed all styles from this PR though, in favor of the styles one :)

File: packages/react-select/src/stories/SelectControlled.stories.tsx
Comment (smhigley) [reply packages/react-select/src/stories/SelectControlled.stories.tsx:5]: @ling1726 here's the issue: https://github.com/microsoft/fluentui/issues/21149

File: packages/react-select/src/components/Select/Select.types.ts
Comment (ling1726) [reply packages/react-select/src/components/Select/Select.types.ts:29]: ah yeah I see 👍

File: packages/react-select/src/components/Select/useSelect.ts
Comment (ling1726) [inline packages/react-select/src/components/Select/useSelect.ts:25]:
```suggestion
    excludedPropNames: ['size'],
```

Are `appearance` and `inline` native attributes ? I can't seem to find any reference documentation about them

**Review by ling1726 — APPROVED** [review]

File: packages/react-select/.storybook/tsconfig.json
Comment (layershifter) [inline packages/react-select/.storybook/tsconfig.json:7]:
```suggestion
    "types": ["static-assets", "environment", "storybook__addons"]
```

We don't have typings for `inline-style-expand-shorthand`, surprised if it passing CI at all :)

File: packages/react-select/jest.config.js
Comment (layershifter) [inline packages/react-select/jest.config.js:20]:
```suggestion
  snapshotSerializers: ['@griffel/jest-serializer'],
```

Should use Griffel

File: packages/react-select/package.json
Comment (layershifter) [inline packages/react-select/package.json:51]:
```suggestion
    "@fluentui/react-utilities": "9.0.0-beta.4",
    "@griffel/react": "1.0.0",
    "tslib": "^2.1.0"
```

File: packages/react-select/package.json
Comment (layershifter) [inline packages/react-select/package.json:31]:
```suggestion
```

File: packages/react-select/tsconfig.lib.json
Comment (layershifter) [inline packages/react-select/tsconfig.lib.json:8]:
```suggestion
    "types": ["static-assets", "environment"]
```

**Review by layershifter — COMMENTED** [review]
Please merge `master`, I also dropped comments, this should help to pass CI on this branch 🐱

File: packages/react-select/src/components/Select/Select.types.ts
Comment (smhigley) [reply packages/react-select/src/components/Select/Select.types.ts:30]: added!

**Comment by smhigley** [comment]
@layershifter thanks for the review! Everything you mentioned should be updated now -- I pulled in the latest master, and updated the outdated patterns :)

**Review by layershifter — APPROVED** [review]

File: packages/react-select/.babelrc.json
Comment (layershifter) [inline packages/react-select/.babelrc.json:2]:
```suggestion
  "presets": ["@griffel"],
  "plugins": ["annotate-pure-calls", "@babel/transform-react-pure-annotations"]
```

This should use Griffel preset :)

**Review by layershifter — APPROVED** [review]
One minor change, other than that we are good to go 👍

File: packages/react-select/src/components/Select/Select.types.ts
Comment (micahgodbolt) [reply packages/react-select/src/components/Select/Select.types.ts:30]: is the inline prop consistent throughout all inputs? as something that you could easily do with CSS for this to make sense, we should have it on every control and added to conformance tests.
I never added this to the slider. Should I have?

File: packages/react-select/src/components/Select/Select.types.ts
Comment (micahgodbolt) [inline packages/react-select/src/components/Select/Select.types.ts:36]: how does darker and lighter work with high contrast or dark themes? Is there a need for color specific appearances?

**Review by micahgodbolt — COMMENTED** [review]
Few prop related questions, but everything else looks good

File: packages/react-select/src/components/Select/Select.types.ts
Comment (smhigley) [reply packages/react-select/src/components/Select/Select.types.ts:36]: They don't affect high contrast, but there are designs in the Figma spec for them in dark theme :). The appearance prop and what it controls is pretty much the exact same as for Input.

Comment (micahgodbolt) [reply packages/react-select/src/components/Select/Select.types.ts:36]: just worried about a future where the design changes a bit and 'darker' is no longer 'darker'. Curious if there was a discussion on how we named those variants in our system? Obviously it seems that it's used a bunch across different components, so would want to ensure consistency across all of them.

**Pull request merged**
